### PR TITLE
add --skip-install arg to create-next-app

### LINKED ISSUE #32367: --skip-install option in create-next-app
### Describe the feature you'd like to request

A `--skip-install` option in create-next-app, like in react-native, to skip installing dependencies would be really great. It allows using other package managers like `pnpm`

### Describe the solution you'd like

do not install dependencies automatically when the `--skip-install` option is set. But the dependencies should be listed in `package.json`, so user could manually install the dependencies using other package managers like `pnpm`

### Describe alternatives you've considered

An option to use `pnpm`(`--use-pnpm`). `pnpm` has gained significant popularity and would be a good option to add.

### PR DESCRIPTION
**Pull request by darshkpatel**
Fixes: #32367

This is an iteration over #32484, which was stalled due to tests not being added/Review comments not being looked into for a while so I took up the initiative to fix up the PR and add the necessary tests. I've cherry-picked @Pika-Pool 's commits to attribute his contributions. 

This PR adds the `--skip-install` feature for create-next-app discussed in #32367 facilitating the use of any package manager (cough cough, pnpm) the user wishes to utilize for installing the dependencies while using create-next-app.

Examples already have a package.json file in the directory, but templates don't. Defaulting to "latest" dist tag for these dependencies as discussed in https://github.com/vercel/next.js/issues/32367#issuecomment-993054819 .

## Feature

- [x] Implements an existing feature request or RFC. Make sure the feature request has been accepted for implementation before opening a PR.
- [x] Related issues linked using `fixes #number`
- [x] Integration tests added
- [x] Documentation added

## Documentation / Examples

- [x] Make sure the linting passes by running `yarn lint`

### CODE REVIEW

**Comment by ijjk** [comment]
# Stats from current PR

<details>
<summary><strong>Default Build</strong> (Increase detected ⚠️)</summary>

<br/>

<details>
<summary><strong>General</strong> Overall increase ⚠️</summary>

|  | vercel/next.js canary  | darshkpatel/next.js skipInstall | Change |
| - | - | - | - |
| buildDuration | 13.9s | 14.1s | ⚠️ +118ms |
| buildDurationCached | 3s | 3.1s | ⚠️ +78ms |
| nodeModulesSize | 348 MB | 348 MB | ⚠️ +3 B |

</details>

<details>
<summary><strong>Page Load Tests</strong> Overall increase ✓</summary>

|  | vercel/next.js canary  | darshkpatel/next.js skipInstall | Change |
| - | - | - | - |
| / failed reqs | 0 | 0 | ✓ |
| / total time (seconds) | 2.861 | 2.723 | -0.14 |
| / avg req/sec | 873.79 | 917.95 | +44.16 |
| /error-in-render failed reqs | 0 | 0 | ✓ |
| /error-in-render total time (seconds) | 1.333 | 1.328 | 0 |
| /error-in-render avg req/sec | 1875.23 | 1882.56 | +7.33 |

</details>

<details>
<summary><strong>Client Bundles (main, webpack, commons)</strong></summary>

|  | vercel/next.js canary  | darshkpatel/next.js skipInstall | Change |
| - | - | - | - |
| 450.HASH.js gzip | 179 B | 179 B | ✓ |
| framework-HASH.js gzip | 42.2 kB | 42.2 kB | ✓ |
| main-HASH.js gzip | 27.1 kB | 27.1 kB | ✓ |
| webpack-HASH.js gzip | 1.45 kB | 1.45 kB | ✓ |
| Overall change | 70.9 kB | 70.9 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | darshkpatel/next.js skipInstall | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 31 kB | 31 kB | ✓ |
| Overall change | 31 kB | 31 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | vercel/next.js canary  | darshkpatel/next.js skipInstall | Change |
| - | - | - | - |
| _app-HASH.js gzip | 1.37 kB | 1.37 kB | ✓ |
| _error-HASH.js gzip | 194 B | 194 B | ✓ |
| amp-HASH.js gzip | 312 B | 312 B | ✓ |
| css-HASH.js gzip | 326 B | 326 B | ✓ |
| dynamic-HASH.js gzip | 2.37 kB | 2.37 kB | ✓ |
| head-HASH.js gzip | 350 B | 350 B | ✓ |
| hooks-HASH.js gzip | 919 B | 919 B | ✓ |
| image-HASH.js gzip | 4.74 kB | 4.74 kB | ✓ |
| index-HASH.js gzip | 263 B | 263 B | ✓ |
| link-HASH.js gzip | 2.13 kB | 2.13 kB | ✓ |
| routerDirect..HASH.js gzip | 321 B | 321 B | ✓ |
| script-HASH.js gzip | 383 B | 383 B | ✓ |
| withRouter-HASH.js gzip | 318 B | 318 B | ✓ |
| 85e02e95b279..7e3.css gzip | 107 B | 107 B | ✓ |
| Overall change | 14.1 kB | 14.1 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | darshkpatel/next.js skipInstall | Change |
| - | - | - | - |
| _buildManifest.js gzip | 459 B | 459 B | ✓ |
| Overall change | 459 B | 459 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | vercel/next.js canary  | darshkpatel/next.js skipInstall | Change |
| - | - | - | - |
| index.html gzip | 532 B | 532 B | ✓ |
| link.html gzip | 546 B | 546 B | ✓ |
| withRouter.html gzip | 527 B | 527 B | ✓ |
| Overall change | 1.6 kB | 1.6 kB | ✓ |

</details>

</details>
<hr/>
<details>
<summary><strong>Default Build with SWC</strong> (Increase detected ⚠️)</summary>

<br/>

<details>
<summary><strong>General</strong> Overall increase ⚠️</summary>

|  | vercel/next.js canary  | darshkpatel/next.js skipInstall | Change |
| - | - | - | - |
| buildDuration | 15.4s | 14.9s | -476ms |
| buildDurationCached | 3.2s | 3.2s | ⚠️ +6ms |
| nodeModulesSize | 348 MB | 348 MB | ⚠️ +3 B |

</details>

<details>
<summary><strong>Page Load Tests</strong> Overall increase ✓</summary>

|  | vercel/next.js canary  | darshkpatel/next.js skipInstall | Change |
| - | - | - | - |
| / failed reqs | 0 | 0 | ✓ |
| / total time (seconds) | 2.908 | 2.876 | -0.03 |
| / avg req/sec | 859.81 | 869.23 | +9.42 |
| /error-in-render failed reqs | 0 | 0 | ✓ |
| /error-in-render total time (seconds) | 1.33 | 1.325 | -0.01 |
| /error-in-render avg req/sec | 1879.68 | 1887.28 | +7.6 |

</details>

<details>
<summary><strong>Client Bundles (main, webpack, commons)</strong></summary>

|  | vercel/next.js canary  | darshkpatel/next.js skipInstall | Change |
| - | - | - | - |
| 450.HASH.js gzip | 179 B | 179 B | ✓ |
| framework-HASH.js gzip | 42.3 kB | 42.3 kB | ✓ |
| main-HASH.js gzip | 27.2 kB | 27.2 kB | ✓ |
| webpack-HASH.js gzip | 1.44 kB | 1.44 kB | ✓ |
| Overall change | 71.1 kB | 71.1 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | darshkpatel/next.js skipInstall | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 31 kB | 31 kB | ✓ |
| Overall change | 31 kB | 31 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | vercel/next.js canary  | darshkpatel/next.js skipInstall | Change |
| - | - | - | - |
| _app-HASH.js gzip | 1.35 kB | 1.35 kB | ✓ |
| _error-HASH.js gzip | 180 B | 180 B | ✓ |
| amp-HASH.js gzip | 305 B | 305 B | ✓ |
| css-HASH.js gzip | 321 B | 321 B | ✓ |
| dynamic-HASH.js gzip | 2.36 kB | 2.36 kB | ✓ |
| head-HASH.js gzip | 342 B | 342 B | ✓ |
| hooks-HASH.js gzip | 906 B | 906 B | ✓ |
| image-HASH.js gzip | 4.76 kB | 4.76 kB | ✓ |
| index-HASH.js gzip | 256 B | 256 B | ✓ |
| link-HASH.js gzip | 2.19 kB | 2.19 kB | ✓ |
| routerDirect..HASH.js gzip | 314 B | 314 B | ✓ |
| script-HASH.js gzip | 375 B | 375 B | ✓ |
| withRouter-HASH.js gzip | 309 B | 309 B | ✓ |
| 85e02e95b279..7e3.css gzip | 107 B | 107 B | ✓ |
| Overall change | 14.1 kB | 14.1 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | darshkpatel/next.js skipInstall | Change |
| - | - | - | - |
| _buildManifest.js gzip | 459 B | 459 B | ✓ |
| Overall change | 459 B | 459 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | vercel/next.js canary  | darshkpatel/next.js skipInstall | Change |
| - | - | - | - |
| index.html gzip | 531 B | 531 B | ✓ |
| link.html gzip | 546 B | 546 B | ✓ |
| withRouter.html gzip | 526 B | 526 B | ✓ |
| Overall change | 1.6 kB | 1.6 kB | ✓ |

</details>

</details>
Commit: 16cb89158a504c690a5b08373d4a417e73c535cd

**Comment by ijjk** [comment]
# Stats from current PR

<details>
<summary><strong>Default Build</strong></summary>

<br/>

<details>
<summary><strong>General</strong> Overall increase ⚠️</summary>

|  | vercel/next.js canary  | darshkpatel/next.js skipInstall | Change |
| - | - | - | - |
| buildDuration | 13.6s | 13.6s | ⚠️ +9ms |
| buildDurationCached | 3.1s | 3.1s | ⚠️ +5ms |
| nodeModulesSize | 348 MB | 348 MB | ⚠️ +6 B |

</details>

<details>
<summary><strong>Page Load Tests</strong> Overall increase ✓</summary>

|  | vercel/next.js canary  | darshkpatel/next.js skipInstall | Change |
| - | - | - | - |
| / failed reqs | 0 | 0 | ✓ |
| / total time (seconds) | 2.729 | 2.697 | -0.03 |
| / avg req/sec | 916.02 | 926.82 | +10.8 |
| /error-in-render failed reqs | 0 | 0 | ✓ |
| /error-in-render total time (seconds) | 1.151 | 1.154 | 0 |
| /error-in-render avg req/sec | 2171.87 | 2165.65 | ⚠️ -6.22 |

</details>

<details>
<summary><strong>Client Bundles (main, webpack, commons)</strong></summary>

|  | vercel/next.js canary  | darshkpatel/next.js skipInstall | Change |
| - | - | - | - |
| 450.HASH.js gzip | 179 B | 179 B | ✓ |
| framework-HASH.js gzip | 42.2 kB | 42.2 kB | ✓ |
| main-HASH.js gzip | 27.1 kB | 27.1 kB | ✓ |
| webpack-HASH.js gzip | 1.45 kB | 1.45 kB | ✓ |
| Overall change | 70.9 kB | 70.9 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | darshkpatel/next.js skipInstall | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 31 kB | 31 kB | ✓ |
| Overall change | 31 kB | 31 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | vercel/next.js canary  | darshkpatel/next.js skipInstall | Change |
| - | - | - | - |
| _app-HASH.js gzip | 1.37 kB | 1.37 kB | ✓ |
| _error-HASH.js gzip | 194 B | 194 B | ✓ |
| amp-HASH.js gzip | 312 B | 312 B | ✓ |
| css-HASH.js gzip | 326 B | 326 B | ✓ |
| dynamic-HASH.js gzip | 2.37 kB | 2.37 kB | ✓ |
| head-HASH.js gzip | 350 B | 350 B | ✓ |
| hooks-HASH.js gzip | 919 B | 919 B | ✓ |
| image-HASH.js gzip | 4.74 kB | 4.74 kB | ✓ |
| index-HASH.js gzip | 263 B | 263 B | ✓ |
| link-HASH.js gzip | 2.13 kB | 2.13 kB | ✓ |
| routerDirect..HASH.js gzip | 321 B | 321 B | ✓ |
| script-HASH.js gzip | 383 B | 383 B | ✓ |
| withRouter-HASH.js gzip | 318 B | 318 B | ✓ |
| 85e02e95b279..7e3.css gzip | 107 B | 107 B | ✓ |
| Overall change | 14.1 kB | 14.1 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | darshkpatel/next.js skipInstall | Change |
| - | - | - | - |
| _buildManifest.js gzip | 459 B | 459 B | ✓ |
| Overall change | 459 B | 459 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | vercel/next.js canary  | darshkpatel/next.js skipInstall | Change |
| - | - | - | - |
| index.html gzip | 532 B | 532 B | ✓ |
| link.html gzip | 546 B | 546 B | ✓ |
| withRouter.html gzip | 527 B | 527 B | ✓ |
| Overall change | 1.6 kB | 1.6 kB | ✓ |

</details>

</details>
<hr/>
<details>
<summary><strong>Default Build with SWC</strong> (Increase detected ⚠️)</summary>

<br/>

<details>
<summary><strong>General</strong> Overall increase ⚠️</summary>

|  | vercel/next.js canary  | darshkpatel/next.js skipInstall | Change |
| - | - | - | - |
| buildDuration | 15s | 15.1s | ⚠️ +43ms |
| buildDurationCached | 3.1s | 3.1s | ⚠️ +2ms |
| nodeModulesSize | 348 MB | 348 MB | ⚠️ +6 B |

</details>

<details>
<summary><strong>Page Load Tests</strong> Overall increase ✓</summary>

|  | vercel/next.js canary  | darshkpatel/next.js skipInstall | Change |
| - | - | - | - |
| / failed reqs | 0 | 0 | ✓ |
| / total time (seconds) | 2.702 | 2.674 | -0.03 |
| / avg req/sec | 925.3 | 935.03 | +9.73 |
| /error-in-render failed reqs | 0 | 0 | ✓ |
| /error-in-render total time (seconds) | 1.143 | 1.137 | -0.01 |
| /error-in-render avg req/sec | 2187.86 | 2199.02 | +11.16 |

</details>

<details>
<summary><strong>Client Bundles (main, webpack, commons)</strong></summary>

|  | vercel/next.js canary  | darshkpatel/next.js skipInstall | Change |
| - | - | - | - |
| 450.HASH.js gzip | 179 B | 179 B | ✓ |
| framework-HASH.js gzip | 42.3 kB | 42.3 kB | ✓ |
| main-HASH.js gzip | 27.2 kB | 27.2 kB | ✓ |
| webpack-HASH.js gzip | 1.44 kB | 1.44 kB | ✓ |
| Overall change | 71.1 kB | 71.1 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | darshkpatel/next.js skipInstall | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 31 kB | 31 kB | ✓ |
| Overall change | 31 kB | 31 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | vercel/next.js canary  | darshkpatel/next.js skipInstall | Change |
| - | - | - | - |
| _app-HASH.js gzip | 1.35 kB | 1.35 kB | ✓ |
| _error-HASH.js gzip | 180 B | 180 B | ✓ |
| amp-HASH.js gzip | 305 B | 305 B | ✓ |
| css-HASH.js gzip | 321 B | 321 B | ✓ |
| dynamic-HASH.js gzip | 2.36 kB | 2.36 kB | ✓ |
| head-HASH.js gzip | 342 B | 342 B | ✓ |
| hooks-HASH.js gzip | 906 B | 906 B | ✓ |
| image-HASH.js gzip | 4.76 kB | 4.76 kB | ✓ |
| index-HASH.js gzip | 256 B | 256 B | ✓ |
| link-HASH.js gzip | 2.19 kB | 2.19 kB | ✓ |
| routerDirect..HASH.js gzip | 314 B | 314 B | ✓ |
| script-HASH.js gzip | 375 B | 375 B | ✓ |
| withRouter-HASH.js gzip | 309 B | 309 B | ✓ |
| 85e02e95b279..7e3.css gzip | 107 B | 107 B | ✓ |
| Overall change | 14.1 kB | 14.1 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | darshkpatel/next.js skipInstall | Change |
| - | - | - | - |
| _buildManifest.js gzip | 459 B | 459 B | ✓ |
| Overall change | 459 B | 459 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | vercel/next.js canary  | darshkpatel/next.js skipInstall | Change |
| - | - | - | - |
| index.html gzip | 531 B | 531 B | ✓ |
| link.html gzip | 546 B | 546 B | ✓ |
| withRouter.html gzip | 526 B | 526 B | ✓ |
| Overall change | 1.6 kB | 1.6 kB | ✓ |

</details>

</details>
Commit: 91bf187497cd13e0edfa624cc486e74e0d4a2076

File: packages/create-next-app/README.md [hunk 1]
```diff
@@ -39,6 +39,7 @@ pnpm create next-app blog-app
 - **--example-path &lt;path-to-example&gt;** - In a rare case, your GitHub URL might contain a branch name with a slash (e.g. bug/fix-1) and the path to the example (e.g. foo/bar). In this case, you must specify the path to the example separately: `--example-path foo/bar`
 - **--use-npm** - Explicitly tell the CLI to bootstrap the app using npm. To bootstrap using yarn we recommend to run `yarn create next-app`
 - **--use-pnpm** - Explicitly tell the CLI to bootstrap the app using pnpm. To bootstrap using yarn we recommend running `yarn create next-app`
+- **--skip-install** - Skips dependency installation step. You can use your favorite package manager.
```
Comment (styfle) [inline packages/create-next-app/README.md:42]: I'm thinking this should be `--no-install`.

Also, is it still needed now that we have `--use-pnpm` which satisfies #32367?

Comment (darshkpatel) [reply packages/create-next-app/README.md:42]: IMHO, It's better than adding a flag for each major package manager.

**Comment by ijjk** [comment]
## Failing test suites 
Commit: dd445e322b4f79bf0209ce40d83c92acdbe4c4f1

`yarn testheadless test/integration/create-next-app/index.test.ts` 
- create next app > should skip installation on supplying --no-install
- create next app > should generate package.json on supplying --no-install
- create next app > should generate package.json with ts dependencies on supplying --no-install and --typescript
- create next app > should skip installation on supplying --no-install with example

<details>
<summary>Expand output</summary>

  ● create next app › should skip installation on supplying --no-install

    expect(received).toBeFalsy()

    Received: true

      473 |       expect(
      474 |         fs.existsSync(path.join(cwd, projectName, 'node_modules'))
    > 475 |       ).toBeFalsy()
          |         ^
      476 |     })
      477 |   })
      478 |

      at integration/create-next-app/index.test.ts:475:9
          at runMicrotasks (<anonymous>)
      at usingTempDir (integration/create-next-app/index.test.ts:19:5)
      at Object.<anonymous> (integration/create-next-app/index.test.ts:462:5)

  ● create next app › should generate package.json on supplying --no-install

    expect(received).toMatchObject(expected)

    - Expected  - 3
    + Received  + 3

      Object {
    -   "next": "latest",
    -   "react": "latest",
    -   "react-dom": "latest",
    +   "next": "12.1.5",
    +   "react": "18.0.0",
    +   "react-dom": "18.0.0",
      }

      488 |       // Assert for default dependencies
      489 |       const pkgJSON = require(pkgJSONPath)
    > 490 |       expect(pkgJSON.dependencies).toMatchObject({
          |                                    ^
      491 |         react: 'latest',
      492 |         'react-dom': 'latest',
      493 |         next: 'latest',

      at integration/create-next-app/index.test.ts:490:36
          at runMicrotasks (<anonymous>)
      at usingTempDir (integration/create-next-app/index.test.ts:19:5)
      at Object.<anonymous> (integration/create-next-app/index.test.ts:480:5)

  ● create next app › should generate package.json with ts dependencies on supplying --no-install and --typescript

    expect(received).toMatchObject(expected)

    - Expected  - 3
    + Received  + 3

      Object {
    -   "next": "latest",
    -   "react": "latest",
    -   "react-dom": "latest",
    +   "next": "12.1.5",
    +   "react": "18.0.0",
    +   "react-dom": "18.0.0",
      }

      513 |       // Assert for default dependencies for ts projects
      514 |       const pkgJSON = require(pkgJSONPath)
    > 515 |       expect(pkgJSON.dependencies).toMatchObject({
          |                                    ^
      516 |         react: 'latest',
      517 |         'react-dom': 'latest',
      518 |         next: 'latest',

      at integration/create-next-app/index.test.ts:515:36
          at runMicrotasks (<anonymous>)
      at usingTempDir (integration/create-next-app/index.test.ts:19:5)
      at Object.<anonymous> (integration/create-next-app/index.test.ts:503:5)

  ● create next app › should skip installation on supplying --no-install with example

    expect(received).toBeFalsy()

    Received: true

      550 |       expect(
      551 |         fs.existsSync(path.join(cwd, projectName, 'node_modules'))
    > 552 |       ).toBeFalsy()
          |         ^
      553 |     })
      554 |   })
      555 | })

      at integration/create-next-app/index.test.ts:552:9
          at runMicrotasks (<anonymous>)
      at usingTempDir (integration/create-next-app/index.test.ts:19:5)
      at Object.<anonymous> (integration/create-next-app/index.test.ts:531:5)

</details>


Read more about building and testing Next.js in [contributing.md](https://github.com/vercel/next.js/blob/canary/contributing.md).

**Comment by darshkpatel** [comment]
The package [commander](https://www.npmjs.com/package/commander#common-option-types-boolean-and-value) changes --no-install to install with the value false. Just another reminder to read the documentation properly 😃 

Will push a fix

**Comment by balazsorban44** [comment]
So since we added support for `--use-pnpm`, and `--use-npm` was already supported (yarn used if neither of the flags passed), I wonder if there is still a use case for `--skip-install`.

I believe the first thing one does after creating a repo is to install its dependencies.


https://github.com/vercel/next.js/issues/32367#issue-1076813859 mentioned `--use-pnpm` as a good alternative for their feature request.

Can we close this?

**Comment by HassanAlthaf** [comment]
@balazsorban44 there is definitely a use-case for this when using monorepos. Please refer to this video to understand in depth. Skip to 14:40s.

**Comment by balazsorban44** [comment]
If you run the install command in a workspace, I believe your package manager should be smart enough to hoist dependencies to the root and symlink them inside the project directory's `node_modules` folder.

I just tried with both pnpm and yarn and there were no issues running the `create-next-app` CLI in a monorepo.

**Pull request closed without merging**